DMD-452 - load workspace copy im

vojtabiberle opened the pull request:
Linear: [DMD-452](https://linear.app/keboola/issue/DMD-452/bq-copy-im)
Connection PR: https://github.com/keboola/connection/pull/6455

---

## Summary
Implements `LoadTableToWorkspaceHandler` for BigQuery to support loading tables to workspaces with COPY import mode. Also includes fixes for CI test stability issues.

## Changes
- Added `LoadTableToWorkspaceHandler` to handle workspace table loading operations
- Updated `keboola/storage-driver-common` to version 7.10.0
- Added comprehensive tests for workspace loading from linked buckets
- Added CLAUDE.md documentation file

## Updates since last revision
Fixed race conditions and test stability issues causing intermittent CI test failures:

1. **Hash collision fix** (`BaseCase.php`): Modified `getTestHash()` to include `static::class` in the hash input. This prevents hash collisions between different test classes that have methods with the same name (e.g., `testImportTableFromTableIncrementalLoad` exists in both `LoadIncrementalFromTableTest` and `IncrementalImportTableFromTableTest`).

2. **IAM 412 error fix** (`ObjectInfoTest.php`): Added retry logic with exponential backoff for GCS IAM policy updates. When multiple tests run in parallel and update the same bucket's IAM policy, the ETag can change between read and write, causing 412 (Precondition Failed) errors. The fix reloads the policy before each retry attempt.

3. **Code style fixes**: Added proper `use` statement for `ServiceException` class and sorted use statements alphabetically in `CreateDropWorkspaceTest.php`.

4. **ExportTableToFileTest fix**: Updated `testExportTableToSlicedFile` to build expected file paths dynamically using `getTestHash()` instead of hardcoded hash values in the data provider. This was necessary because the `getTestHash()` change affected all hash values.

5. **CreateDropWorkspaceTest fix**: Modified the permission check test to use direct bucket creation (via `CreateBucketHandler`) with a unique bucket name instead of calling `createTestBucket()`. The previous approach included a cleanup step that could fail silently with low-privilege workspace credentials, causing 409 (conflict) errors instead of the expected 403 (permission denied).

## Human Review Checklist
- [ ] Verify `getTestHash()` change doesn't break tests that may rely on stable bucket names across test methods
- [ ] Verify IAM retry logic properly reloads policy before each retry attempt
- [ ] Verify ExportTableToFileTest still properly validates export functionality with dynamic paths
- [ ] Verify CreateDropWorkspaceTest permission check correctly tests authorization (not just conflict)
- [ ] Confirm LoadTableToWorkspaceHandler implementation matches expected behavior
- [ ] Review removed test cases in `ExportQueryBuilderTest.php` (Iso column and case-insensitive filter tests) - these were removed from the test table columns

---

Link to Devin run: https://app.devin.ai/sessions/4fe69aaff39e4fd2a5e7e99630c784fc
Requested by: vojtech.biberle@keboola.com (@vojtabiberle)

[DMD-452]: https://keboola.atlassian.net/browse/DMD-452?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vojtabiberle committed:
DMD-452 - handler copy + tests copy

vojtabiberle committed:
Revert "Merge pull request #183 from keboola/vb/DMD-452/bigquery-copy-im-II"

This reverts commit 4e5d5fc3ea220776b2eeb61492a0d7c5444c7fde, reversing
changes made to 64ec290cb295449ed73f7ec0f6fb3142dcf1ee39.

vojtabiberle committed:
DMD-452 - COPY IM in LoadTableToWorkspace

vojtabiberle committed:
DMD-452 - COPY IM tests

vojtabiberle committed:
DMD-452 - Claude init

vojtabiberle committed:
DMD-452 - fix codestyle

vojtabiberle committed:
DMD-452 - fix hardcoded bucket name in tests

jirkasemmler commented on src/Handler/Table/Import/ImportTableFromTableHandler.php:
```diff
@@ -48,21 +47,11 @@
 final class ImportTableFromTableHandler extends BaseHandler
 {
     public GCPClientManager $clientManager;
-    private ?ImportSourceFactory $sourceFactory;
```
jen poznamka : zkontroloval jsem, ze v teto podobe je to shodne se stavem "pred vojtovou prvni upravou", tzn s revizi https://github.com/keboola/php-storage-driver-bigquery/pull/184

jirkasemmler commented on src/Handler/Workspace/Load/LoadTableToWorkspaceHandler.php:
```diff
@@ -0,0 +1,476 @@
+     * @throws ObjectAlreadyExistsException
+     * @throws ConflictException
+     */
+    private function createView(
```
valuduju, ze to je copypaste z ImportTableFromTableHandlaru

vojtabiberle replied:
Jak jsem psal ve Slacku, pustil jsem na to i Claude, takže tam možná malé změny budou, ještě jsem to sám nestihl detailně projít

jirkasemmler replied:
proto to validuju. jen teda bych to mohl poslat po jednom, priznavam

jirkasemmler commented on src/Handler/Workspace/Load/LoadTableToWorkspaceLib/CopyImportFromTableToTable.php:
```diff
@@ -0,0 +1,85 @@
+use Keboola\TableBackendUtils\Table\Bigquery\BigqueryTableReflection;
+use Keboola\TableBackendUtils\Table\TableDefinitionInterface;
+
+class CopyImportFromTableToTable implements ToStageImporterInterface
```
shodne, jen se meni namespace

jirkasemmler commented on src/Handler/Workspace/Load/StageTableDefinitionFactory.php:
```diff
@@ -0,0 +1,114 @@
+use Keboola\TableBackendUtils\Column\ColumnCollection;
+use Keboola\TableBackendUtils\Table\Bigquery\BigqueryTableDefinition;
+
+final class StageTableDefinitionFactory
```
shodne s table import

jirkasemmler commented on tests/functional/UseCase/Workspace/Load/LoadTableFromTableTest.php:
```diff
@@ -0,0 +1,3619 @@
+        $this->assertSame($ref->getRowsCount(), $response->getTableRowsCount());
+
+        // cleanup
+        $bqClient->runQuery($bqClient->query(
```
hele jak minule, toto nebude stabilni proti padu, ma se to cistit predem

vojtabiberle replied:
fixed

jirkasemmler commented on tests/functional/UseCase/Workspace/Load/LoadTableFromTableTest.php:
```diff
@@ -0,0 +1,3619 @@
+            ]),
+            [],
+        );
+        $sql = $qb->getCreateTableCommand(
```
tady to vyrabeni cilove tabulky je zbytecne, ale existuje testCopyCreatesDestinationTableWhenMissing tak to nechme

vojtabiberle replied:
OK

jirkasemmler commented on tests/functional/UseCase/Workspace/Load/LoadTableFromTableTest.php:
```diff
@@ -0,0 +1,3619 @@
+    /**
+     * @dataProvider typedTablesProvider
+     */
+    public function testIncrementalImportWithNullInDedupColumn(bool $isTypedTable): void
```
proc jsou ty inc testy tady?

vojtabiberle replied:
protože testy delal celý Claude a já je ještě nestihl poopravovat

vojtabiberle replied:
moved

vojtabiberle committed:
DMD-452 - fix test name uniqnes

jirkasemmler reviewed (changes requested):
za me je potreba presunout ty cleanupy v tech testech a jinak ok. Ale mame problem - to COPY vubec neresi castovani dat. A on to neresi ani common driver, ktery jen mapuje nazvy sloupcu. Ale pojdme poporade. dodame toto a pak doresime zbytek

vojtabiberle committed:
DMD-452 - tests cleanup

zajca commented on src/Handler/Workspace/Load/LoadTableToWorkspaceHandler.php:
```diff
@@ -0,0 +1,444 @@
+
+        // Replace is only available in view or clone import
+        $shouldDropTableIfExists = $importOptions->getCreateMode() === ImportOptions\CreateMode::REPLACE
+            && in_array($importOptions->getImportType(), [ImportType::VIEW, ImportType::PBCLONE], true);
```
Toto bych řekl že bude v E2E storage padat při load overwrite, overwrite funguje aj pro copy load.

zajca replied:
Hmm ono sa to dropuje v Connetion zvlášt :thinking:

vojtabiberle replied:
tu nevím, na co narážíš, ale asi je to v pohodě …

zajca commented on src/Handler/Workspace/Load/LoadTableToWorkspaceHandler.php:
```diff
@@ -91,10 +99,40 @@ public function __invoke(
+        // Instantiate services if not injected (for backward compatibility)
+        $sourceFactory = $this->sourceFactory ?? new LoadSourceFactory($bqClient);
+        $columnMapping = $this->columnMappingService ?? new ColumnMappingService();
+        $destinationManager = $this->destinationManager ?? new LoadDestinationManager($bqClient);
```
Tady to vůbec nechápu proč tam je žádné DI sa v driveru nepoužívá, takže toto je kec a ten ctor je jen na zmatený pokud to nemá unit testy tak mě to nedává smysl

vojtabiberle replied:
unit testy to má, je to jen rozdělení moc velké třídy na menší části - přehlednější

ctor tam udělalo AI "kdyby náhodou" jsme to začali používat lépe … komentář je nesmysl, toho jsem si nevšiml

ale smazal jsem oboje

zajca commented on tests/functional/UseCase/Workspace/Load/LoadTableFromTableTest.php:
```diff
@@ -1106,4 +1112,2508 @@ public function testImportFromTypedTableToStringTable(): void
+            // Expected for string tables - BigQuery catches NULL during query execution
+            $this->assertStringContainsString('null', strtolower($e->getMessage()));
+            $this->assertStringContainsString('id', strtolower($e->getMessage()));
+        } catch (Throwable $e) {
```
Tady by měl byt v těch catch ještě assert na isTypedTable když už je to okomentované

vojtabiberle replied:
přidáno

devin-ai-integration[bot] committed:
DMD-452 - fix race condition in parallel tests

- Include class name in getTestHash() to prevent hash collisions between
  different test classes with the same method name
- Add retry logic for IAM policy updates to handle 412 errors from
  concurrent modifications

Co-Authored-By: vojtech.biberle@keboola.com <vojtech.biberle@gmail.com>

devin-ai-integration[bot] committed:
DMD-452 - fix code style: add use statement for ServiceException

Co-Authored-By: vojtech.biberle@keboola.com <vojtech.biberle@gmail.com>

devin-ai-integration[bot] committed:
DMD-452 - fix ExportTableToFileTest to not use hardcoded hash values

The test was using hardcoded hash values in the data provider which broke
when getTestHash() was changed to include class name. Now the test builds
expected file paths dynamically using getTestHash().

Co-Authored-By: vojtech.biberle@keboola.com <vojtech.biberle@gmail.com>

devin-ai-integration[bot] committed:
DMD-452 - fix CreateDropWorkspaceTest permission check

Use direct bucket creation without cleanup step to ensure we get a clean
403 permission error instead of 409 conflict. The previous approach used
createTestBucket() which includes a cleanup step that can fail silently
with low-privilege workspace credentials.

Co-Authored-By: vojtech.biberle@keboola.com <vojtech.biberle@gmail.com>

zajca commented on tests/functional/UseCase/Info/ObjectInfoTest.php:
```diff
@@ -536,16 +536,30 @@ private function prepareConnectionForExternalBucket(): Connection
+                }
+                throw $e;
+            }
+        }
```
Na retry máme našu retry lib je tu různě použitá třeba v https://github.com/keboola/php-storage-driver-bigquery/blob/main/src/Handler/Workspace/Create/CreateWorkspaceHandler.php#L97-L119 téj bysme sa měli držat než tady těch custom vibe coded retry

vojtabiberle replied:
opraveno po devinovi (já říkal, že jsem to ještě nestihl zkouknout a ofiko pustit na review)

zajca commented on tests/functional/BaseCase.php:
```diff
@@ -361,11 +362,46 @@ private function dropTestBucket(
+                ));
+                usleep(500000); // 0.5 second before retry
+            }
+        }
```
Na retry máme našu retry lib je tu různě použitá třeba v https://github.com/keboola/php-storage-driver-bigquery/blob/main/src/Handler/Workspace/Create/CreateWorkspaceHandler.php#L97-L119 téj bysme sa měli držat než tady těch custom vibe coded retry

vojtabiberle replied:
opraveno po devinovi (já říkal, že jsem to ještě nestihl zkouknout a ofiko pustit na review)

devin-ai-integration[bot] committed:
DMD-452 - fix code style: sort use statements alphabetically

Co-Authored-By: vojtech.biberle@keboola.com <vojtech.biberle@gmail.com>

zajca reviewed (approved):
Nenašel sem nic závažného přijde mě že ty testy sů v pohodě všecky sem jich nějak prošel.

vojtabiberle committed:
DMD-452 - move tests

vojtabiberle committed:
DMD-452 - remove unused ctor

vojtabiberle committed:
DMD-452 - assert for typed table

vojtabiberle committed:
DMD-452 - fix retry
